library/x509_crt.c

```diff
@@ -742,270 +753,510 @@ static int x509_get_crt_ext( unsigned char **p,
     return( ret );
 }
 
-/*
- * Parse and fill a single X.509 certificate in DER format
- */
-static int x509_crt_parse_der_core( mbedtls_x509_crt *crt,
-                                    const unsigned char *buf,
-                                    size_t buflen,
-                                    int make_copy )
+static int x509_crt_parse_frame( unsigned char *start,
+                                 unsigned char *end,
+                                 mbedtls_x509_crt_frame *frame )
 {
     int ret;
+    unsigned char *p;
     size_t len;
-    unsigned char *p, *end, *crt_end;
-    mbedtls_x509_buf sig_params1, sig_params2, sig_oid2;
 
-    memset( &sig_params1, 0, sizeof( mbedtls_x509_buf ) );
-    memset( &sig_params2, 0, sizeof( mbedtls_x509_buf ) );
-    memset( &sig_oid2, 0, sizeof( mbedtls_x509_buf ) );
+    mbedtls_x509_buf tmp;
+    unsigned char *tbs_start;
 
-    /*
-     * Check for valid input
-     */
-    if( crt == NULL || buf == NULL )
-        return( MBEDTLS_ERR_X509_BAD_INPUT_DATA );
+    mbedtls_x509_buf outer_sig_alg;
+    size_t inner_sig_alg_len;
+    unsigned char *inner_sig_alg_start;
 
-    /* Use the original buffer until we figure out actual length. */
-    p = (unsigned char*) buf;
-    len = buflen;
-    end = p + len;
+    memset( frame, 0, sizeof( *frame ) );
 
     /*
-     * Certificate  ::=  SEQUENCE  {
+     * Certificate  ::=  SEQUENCE {
      *      tbsCertificate       TBSCertificate,
      *      signatureAlgorithm   AlgorithmIdentifier,
-     *      signatureValue       BIT STRING  }
+     *      signatureValue       BIT STRING
+     * }
+     *
      */
+    p = start;
+
+    frame->raw.p = p;
     if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
             MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) != 0 )
     {
-        mbedtls_x509_crt_free( crt );
         return( MBEDTLS_ERR_X509_INVALID_FORMAT );
     }
 
-    end = crt_end = p + len;
-    crt->raw.len = crt_end - buf;
-    if( make_copy != 0 )
+    /* NOTE: We are currently not checking that the `Certificate`
+     * structure spans the entire buffer. */
```

Comment:
No, we're not doing this, because IMHO some users rely on it, even though it's wrong - let's follow up internally.

Regarding the comment: The check to which it applies ensures that the `Certificate` structure does not contain additional bytes after the `TBSCertificate` and the Signature. The check we deliberately omit here is that the input buffer doesn't contain additional bytes after the `Certificate` structure.

Reply:
OK, this is a deliberate omission, not a `TODO` for us to come add it later.

Reply:
No change.